File: ios/RePaperKit/Sources/RePaperKit/OdDevice.swift
import Foundation

/// One BLE link to a sheet, transport-agnostic: CoreBluetooth implements OdLink in the app;
/// tests script it. Auth, interrogate, legacy uncompressed direct-write upload with per-chunk ACKs.
public protocol OdLink: Sendable {
    func write(_ data: Data, withResponse: Bool) async throws
    func read(timeoutMs: Int) async throws -> Data
}

public actor OdDevice {
    let link: OdLink
    let masterKey: Data?
    var sessionKey: Data?
    var sessionId: Data?
    var nonceCounter: UInt64 = 0
    public private(set) var capabilities: Capabilities?
    public private(set) var lastConfigHex: String?

    public static let timeoutAck = 5_000
    public static let timeoutFirstChunk = 10_000
    public static let timeoutConfigChunk = 2_000
    public static let timeoutDataAck = 90_000
    public static let timeoutRefresh = 90_000

    public init(link: OdLink, masterKey: Data? = nil) {
        self.link = link; self.masterKey = masterKey
    }

    public var isAuthenticated: Bool { sessionKey != nil }

    /// Test hook: preload capabilities as if interrogate had run.
    public func _setCapabilitiesForTesting(_ c: Capabilities) { capabilities = c }

    public func authenticate() async throws {
        guard let key = masterKey else { throw OdError.authError("sheet requires a key but none is registered") }
        var challenge: (Data, Data)?
        for attempt in 0..<2 {
            try await link.write(Od.authStep1(), withResponse: true)
            do { challenge = try Od.parseAuthChallenge(try await link.read(timeoutMs: Self.timeoutAck)); break }
            catch OdError.authError(let m) where m.contains("active session") && attempt == 0 { continue }
        }
        guard let (serverNonce, deviceId) = challenge else { throw OdError.authError("no auth challenge") }
        let clientNonce = OdCrypto.clientNonce()
        try await link.write(Od.authStep2(clientNonce: clientNonce,
            challenge: OdCrypto.challengeResponse(master: key, serverNonce: serverNonce, clientNonce: clientNonce, deviceId: deviceId)), withResponse: true)
        let proof = try Od.parseAuthSuccess(try await link.read(timeoutMs: Self.timeoutAck))
        let sk = OdCrypto.deriveSessionKey(master: key, clientNonce: clientNonce, serverNonce: serverNonce, deviceId: deviceId)
        guard proof == OdCrypto.serverProof(sessionKey: sk, serverNonce: serverNonce, clientNonce: clientNonce, deviceId: deviceId) else {
            throw OdError.authError("device failed mutual authentication")
        }
        sessionKey = sk
        sessionId = OdCrypto.deriveSessionId(sessionKey: sk, clientNonce: clientNonce, serverNonce: serverNonce)
        nonceCounter = 0
    }

    public func interrogate() async throws -> Capabilities {
        try await write(Od.readConfig())
        var resp = try await read(timeoutMs: Self.timeoutFirstChunk)
        if resp.count == 4, resp[resp.startIndex] == 0xFF, Int(resp[resp.startIndex + 1]) == (Od.readConfigCode & 0xFF) {
            throw OdError.protocolError("device has no stored configuration")
        }
        var chunk = Od.stripEcho(resp, expected: Od.readConfigCode)
        guard chunk.count >= 4 else { throw OdError.protocolError("config first chunk too short") }
        let total = Int(chunk[chunk.startIndex + 2]) | Int(chunk[chunk.startIndex + 3]) << 8
        var tlv = Data(chunk.dropFirst(4))
        while tlv.count < total {
            resp = try await read(timeoutMs: Self.timeoutConfigChunk)
            chunk = Od.stripEcho(resp, expected: Od.readConfigCode)
            guard chunk.count > 2 else { throw OdError.protocolError("config read stalled") }
            tlv.append(chunk.dropFirst(2))
        }
        lastConfigHex = tlv.hexLower
        let caps = try OdConfig.parse(tlv)
        capabilities = caps
        return caps
    }

    public func print(_ page: Page, fullRefresh: Bool = true, narrate: @Sendable (String) -> Void = { _ in }) async throws {
        if masterKey != nil && !isAuthenticated { narrate("connecting"); try await authenticate() }
        let caps: Capabilities
        if let c = capabilities { caps = c } else { narrate("reading the sheet"); caps = try await interrogate() }
        guard caps.viewedWidth == page.model.width && caps.viewedHeight == page.model.height else {
            throw OdError.protocolError("sheet shows \(caps.viewedWidth)×\(caps.viewedHeight), page is \(page.model.width)×\(page.model.height) — re-register the sheet")
        }
        let native = OdEncoding.rotateToNative(page.indexes, viewedW: page.model.width, viewedH: page.model.height, rotation: caps.rotation)
        let data = OdEncoding.encode(native, width: caps.width, height: caps.height, scheme: caps.scheme, panelIc: caps.panelIc)

        narrate("connected — sending")
        try await write(Od.directWriteStartUncompressed())
        try Od.validateAck(try await read(timeoutMs: Self.timeoutFirstChunk), expected: Od.directWriteStartCode)

        let chunkSize = isAuthenticated ? Od.encryptedChunkSize : Od.chunkSize
        var sent = 0
        var autoCompleted = false
        while sent < data.count {
            let chunk = Data(data.dropFirst(sent).prefix(chunkSize))
            try await write(Od.directWriteData(chunk), withResponse: false)
            sent += chunk.count
            let resp = try await read(timeoutMs: Self.timeoutDataAck)
            switch try Od.code(resp) & ~Od.ackFlag {
            case Od.directWriteEndCode: autoCompleted = true
            case Od.directWriteDataCode: break
            default: throw OdError.protocolError("unexpected response during upload")
            }
            if autoCompleted { break }
        }

        if !autoCompleted {
            try await write(Od.directWriteEnd(refreshMode: fullRefresh ? 0 : 1))
            try Od.validateAck(try await read(timeoutMs: Self.timeoutDataAck), expected: Od.directWriteEndCode)
        }
        narrate("sheet is refreshing (about 20 s)")
        switch try Od.code(try await read(timeoutMs: Self.timeoutRefresh)) & ~Od.ackFlag {
        case Od.refreshCompleteCode: narrate("printed")
        case Od.refreshTimeoutCode: throw OdError.protocolError("display refresh timed out")
        default: throw OdError.protocolError("unexpected response waiting for refresh")
        }
    }

    func write(_ data: Data, withResponse: Bool = true) async throws {
        if let sk = sessionKey, let sid = sessionId {
            let frame = OdCrypto.encryptCommand(sessionKey: sk, sessionId: sid, counter: nonceCounter,
                                                cmd: Data(data.prefix(2)), payload: Data(data.dropFirst(2)))
            nonceCounter += 1
            try await link.write(frame, withResponse: withResponse)
        } else {
            try await link.write(data, withResponse: withResponse)
        }
    }

    func read(timeoutMs: Int) async throws -> Data {
        let raw = try await link.read(timeoutMs: timeoutMs)
        if let sk = sessionKey, raw.count >= 31 {
            let (cmd, payload) = try OdCrypto.decryptResponse(sessionKey: sk, raw: raw)
            return Data([UInt8(cmd >> 8), UInt8(cmd & 0xFF)]) + payload
        }
        if raw.count == 3 && raw[raw.startIndex + 2] == 0xFE { throw OdError.authError("device requires an encryption key") }
        if raw.count == 3 && raw[raw.startIndex + 2] == 0xFF { throw OdError.protocolError("device rejected command: integrity check failed") }
        return raw
    }
}

File: go/core/src/main/kotlin/net/repaper/go/core/OdDevice.kt
package net.repaper.go.core

/** One BLE link to a sheet, transport-agnostic: Android's BluetoothGatt implements [OdLink];
 *  tests script it. Mirrors py-opendisplay's device flow — auth, interrogate, legacy
 *  uncompressed direct-write upload with per-chunk ACKs. */
interface OdLink {
    /** Write to the single OpenDisplay characteristic. [withResponse]=false requests Write Without
     *  Response (bulk 0x71 chunks); implementations may fall back to write-with-response. */
    suspend fun write(data: ByteArray, withResponse: Boolean = true)

    /** Next notification frame, or throw OdError on timeout. */
    suspend fun read(timeoutMs: Long): ByteArray
}

class OdDevice(private val link: OdLink, private val masterKey: ByteArray? = null) {
    private var sessionKey: ByteArray? = null
    private var sessionId: ByteArray? = null
    private var nonceCounter: Long = 0

    var capabilities: Capabilities? = null; private set
    var lastConfigHex: String? = null; private set   // raw TLV of the last interrogate, for diagnostics

    companion object {
        const val TIMEOUT_ACK = 5_000L
        const val TIMEOUT_FIRST_CHUNK = 10_000L
        const val TIMEOUT_CONFIG_CHUNK = 2_000L
        const val TIMEOUT_DATA_ACK = 90_000L
        const val TIMEOUT_END_ACK = 90_000L
        const val TIMEOUT_REFRESH = 90_000L
        const val TIMEOUT_NFC_COMMIT = 15_000L   // the tag EEPROM commit is slow I2C work
    }

    val isAuthenticated: Boolean get() = sessionKey != null

    /** Two-step challenge-response; afterwards every command is CCM-wrapped. */
    suspend fun authenticate() {
        val key = masterKey ?: throw OdAuthError("sheet requires a key but none is registered")
        var challenge: Pair<ByteArray, ByteArray>? = null
        for (attempt in 0 until 2) {
            link.write(Od.authStep1())
            try { challenge = Od.parseAuthChallenge(link.read(TIMEOUT_ACK)); break }
            catch (e: OdAuthError) {
                if (attempt == 1 || !e.message!!.contains("active session")) throw e
            }
        }
        val (serverNonce, deviceId) = challenge!!
        val clientNonce = OdCrypto.clientNonce()
        link.write(Od.authStep2(clientNonce, OdCrypto.challengeResponse(key, serverNonce, clientNonce, deviceId)))
        val proof = Od.parseAuthSuccess(link.read(TIMEOUT_ACK))
        val sk = OdCrypto.deriveSessionKey(key, clientNonce, serverNonce, deviceId)
        // mutual auth: a device that answered OK without the master key can't produce this CMAC
        if (!proof.contentEquals(OdCrypto.serverProof(sk, serverNonce, clientNonce, deviceId))) {
            throw OdAuthError("device failed mutual authentication")
        }
        sessionKey = sk
        sessionId = OdCrypto.deriveSessionId(sk, clientNonce, serverNonce)
        nonceCounter = 0
    }

    /** Read + parse the device config; remembers capabilities. */
    suspend fun interrogate(): Capabilities {
        write(Od.readConfig())
        var resp = read(TIMEOUT_FIRST_CHUNK)
        if (resp.size == 4 && resp[0] == 0xFF.toByte() && (resp[1].toInt() and 0xFF) == (Od.READ_CONFIG and 0xFF)) {
            throw OdError("device has no stored configuration")
        }
        var chunk = Od.stripEcho(resp, Od.READ_CONFIG)
        if (chunk.size < 4) throw OdError("config first chunk too short")
        val total = (chunk[2].toInt() and 0xFF) or ((chunk[3].toInt() and 0xFF) shl 8)
        val tlv = ArrayList<Byte>(total)
        chunk.drop(4).forEach { tlv.add(it) }
        while (tlv.size < total) {
            resp = read(TIMEOUT_CONFIG_CHUNK)
            chunk = Od.stripEcho(resp, Od.READ_CONFIG)
            if (chunk.size <= 2) throw OdError("config read stalled at ${tlv.size}/$total bytes")
            for (i in 2 until chunk.size) tlv.add(chunk[i])   // skip the 2-byte chunk number
        }
        val bytes = tlv.toByteArray()
        lastConfigHex = bytes.toHexLower()
        return OdConfig.parse(bytes).also { capabilities = it }
    }

    /** Upload an already-rendered page and refresh. [narrate] gets human-readable phases. */
    suspend fun print(page: Page, fullRefresh: Boolean = true, narrate: (String) -> Unit = {}) {
        if (masterKey != null && !isAuthenticated) { narrate("connecting"); authenticate() }
        val caps = capabilities ?: run { narrate("reading the sheet"); interrogate() }
        if (caps.viewedWidth != page.model.width || caps.viewedHeight != page.model.height) {
            throw OdError("sheet shows ${caps.viewedWidth}×${caps.viewedHeight}, page is ${page.model.width}×${page.model.height} — re-register the sheet")
        }
        val native = OdEncoding.rotateToNative(page.indexes, page.model.width, page.model.height, caps.rotation)
        val data = OdEncoding.encode(native, caps.width, caps.height, caps.scheme, caps.panelIc)

        narrate("connected — sending")
        write(Od.directWriteStartUncompressed())
        Od.validateAck(read(TIMEOUT_FIRST_CHUNK), Od.DIRECT_WRITE_START)

        val chunkSize = if (isAuthenticated) Od.ENCRYPTED_CHUNK_SIZE else Od.CHUNK_SIZE
        var sent = 0
        var autoCompleted = false
        while (sent < data.size) {
            val chunk = data.copyOfRange(sent, minOf(sent + chunkSize, data.size))
            write(Od.directWriteData(chunk), withResponse = false)
            sent += chunk.size
            val resp = read(TIMEOUT_DATA_ACK)
            when (Od.code(resp) and Od.ACK_FLAG.inv()) {
                Od.DIRECT_WRITE_END -> { autoCompleted = true; break }   // device buffer full: it refreshes by itself
                Od.DIRECT_WRITE_DATA -> {}
                else -> throw OdError("unexpected response during upload: 0x%04x".format(Od.code(resp)))
            }
        }

        if (!autoCompleted) {
            write(Od.directWriteEnd(if (fullRefresh) 0 else 1))
            Od.validateAck(read(TIMEOUT_END_ACK), Od.DIRECT_WRITE_END)
        }
        narrate("sheet is refreshing (about 20 s)")
        when (Od.code(read(TIMEOUT_REFRESH)) and Od.ACK_FLAG.inv()) {
            Od.REFRESH_COMPLETE -> narrate("printed")
            Od.REFRESH_TIMEOUT -> throw OdError("display refresh timed out (device sent 0x74)")
            else -> throw OdError("unexpected response waiting for refresh")
        }
    }

    /** Write the sheet's OWN NFC tag over BLE: a URI record carrying the landing link,
     *  so tapping the sheet always resolves — some sheets ship with an empty tag.
     *  Older firmware stays silent on the unknown opcode: that surfaces as
     *  "doesn't support" (non-fatal for callers). */
    suspend fun writeNfcUrl(url: String) {
        val payload = url.toByteArray(Charsets.UTF_8)
        if (payload.size > Od.NFC_MAX_TOTAL) throw OdError("landing link too long for the tag")
        var first = true
        suspend fun readNfc(timeoutMs: Long): ByteArray = try {
            read(timeoutMs).also { first = false }
        } catch (e: OdError) {
            if (first) throw OdError("this sheet's firmware doesn't support NFC writing") else throw e
        }
        if (payload.size <= Od.NFC_INLINE_MAX) {
            write(Od.nfcWriteInline(Od.NFC_REC_URI, payload))
            Od.validateNfc(readNfc(TIMEOUT_NFC_COMMIT), Od.NFC_STATUS_WRITE_OK)
            return
        }
        write(Od.nfcWriteStart(Od.NFC_REC_URI, payload.size))
        Od.validateNfc(readNfc(TIMEOUT_ACK), Od.NFC_STATUS_CHUNK_ACK)
        var off = 0
        while (off < payload.size) {
            val chunk = payload.copyOfRange(off, minOf(off + Od.NFC_CHUNK, payload.size))
            write(Od.nfcWriteData(chunk))
            Od.validateNfc(readNfc(TIMEOUT_ACK), Od.NFC_STATUS_CHUNK_ACK)
            off += chunk.size
        }
        write(Od.nfcWriteEnd())
        Od.validateNfc(readNfc(TIMEOUT_NFC_COMMIT), Od.NFC_STATUS_WRITE_OK)
    }

    // ── session-aware write/read ─────────────────────────────────────────────

    private suspend fun write(data: ByteArray, withResponse: Boolean = true) {
        val sk = sessionKey; val sid = sessionId
        if (sk != null && sid != null) {
            link.write(OdCrypto.encryptCommand(sk, sid, nonceCounter++, data.copyOfRange(0, 2), data.copyOfRange(2, data.size)), withResponse)
        } else link.write(data, withResponse)
    }

    private suspend fun read(timeoutMs: Long): ByteArray {
        val raw = link.read(timeoutMs)
        val sk = sessionKey
        // Encrypted responses are ≥31 bytes; short frames (direct-write ACKs, error frames) stay plaintext.
        if (sk != null && raw.size >= 31) {
            val (cmd, payload) = OdCrypto.decryptResponse(sk, raw)
            return byteArrayOf((cmd shr 8).toByte(), (cmd and 0xFF).toByte()) + payload
        }
        if (raw.size == 3 && raw[2] == 0xFE.toByte()) throw OdAuthError("device requires an encryption key")
        if (raw.size == 3 && raw[2] == 0xFF.toByte()) throw OdError("device rejected command: integrity check failed")
        return raw
    }
}
